packages/protocols/generation-queue.go

```diff
@@ -0,0 +1,127 @@
+package protocols
+
+import (
+	"errors"
+	"time"
+
+	"github.com/GenesisKernel/go-genesis/packages/model"
+
+	"github.com/GenesisKernel/go-genesis/packages/conf/syspar"
+	log "github.com/sirupsen/logrus"
+)
+
+// QueueChecker allow check queue to generate current block
+type QueueChecker interface {
+	TimeToGenerate(position int64) (bool, error)
+	NextTime(position int64, t time.Time) (time.Time, error)
+	BlockForTimeExists(t time.Time, nodePosition int) (bool, error)
+	RangesByTime(t time.Time) (start, end time.Time)
+}
+
+var (
+	WrongNodePositionError = errors.New("wrong node position")
+	TimeError              = errors.New("current time before first block")
+	DuplicateBlockError    = errors.New("block for this time interval exists")
+)
+
+type BlockTimeCounter struct {
+	start       time.Time
+	duration    time.Duration
+	numberNodes int
+}
+
+// Queue returns serial queue number for time
+func (btc *BlockTimeCounter) Queue(t time.Time) (int, error) {
```

Comment:
BlockID(t time.Time) ?

Reply:
this func returns serial number of generation time interval, that not equal blockID